### PR TITLE
Add JSON-B customization to OptaPlanner extension

### PR DESCRIPTION
**Pull request by cuijulian**
Jira: https://issues.redhat.com/browse/PLANNER-1880

CI build currently fails due to missing `optaplanner-persistence-jsonb` dependency. This dependency will be provided by `optaplanner-bom` once OptaPlanner version `7.35.0.Final` is released. Any feedback is appreciated!

### CODE REVIEW

File: extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java
Comment (ge0ffrey) [inline extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: This is wierd. Does it work?

It doesn't modify the incoming config parameter, it reassigns it. Does this do anything?

File: integration-tests/pom.xml
Comment (ge0ffrey) [inline integration-tests/pom.xml]: Move this under the optaplanner-jackson module a few lines up - keep em together.

File: extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java
Comment (ge0ffrey) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: The integration test proves that this works. I have no idea how it can work though.

**Review by ge0ffrey — APPROVED** [review]

File: extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java
Comment (ge0ffrey) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: @cuijulian Does the integration test fail if you uncomment this line?

**Comment by geoand** [comment]
Thanks for the contribution @cuijulian! 

When you are done with changes, please squash the commits so we can have a meaningful history :)

File: .github/workflows/ci-actions.yml
Comment (geoand) [inline .github/workflows/ci-actions.yml]: I would say let's add another 5 minutes here since we are adding a new module

File: extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java
Comment (geoand) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: Indeed, this looks quite peculiar...

Comment (cuijulian) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: @ge0ffrey The integration test still passes when I comment out that line. My thought process was to customize the `JsonbConfig` into the `OptaPlannerJsonbConfig` (which contains all the Score adapters) by reassigning it, but it seems to have no effect on the integration test. Alternatively, I could also manually customize the `config` using `withAdapters(insert all OptaPlanner Adapters here)`, but I doubt that would have an effect either. @geoand Any thoughts on why this could be?

Separate question: when I build the module with `-Pnative`, the `NativeImageTest` fails with:
```
[ERROR]   OptaPlannerTestResourceIT>OptaPlannerTestResourceTest.solveWithSolverFactory:26 1 expectation failed.
Response body doesn't match expectation.
Expected: is "{\"entityList\":[{\"value\":\"v1\"},{\"value\":\"v2\"}],\"score\":{\"initScore\":0,\"solutionInitialized\":true,\"score\":0},\"valueList\":[\"v1\",\"v2\"]}"
  Actual: {"entityList":[{"value":"v1"},{"value":"v2"}],"score":{"score":0},"valueList":["v1","v2"]}
```
How come the non native integration test passes, but the native IT gives a differently formatted response?

Comment (geoand) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: @cuijulian I am not sure I understand you issue with the `JsonbConfigCustomizer`, can you elaborate please?

As for the test issue, I am not really sure. I would need to look into it, it could be a number of things.
Remind me to look at it tomorrow please as I will surely forget :)

Comment (cuijulian) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: @geoand Basically customizing the `JsonbConfig` config parameter with `JsonbConfigCustomizer` has no effect on the `OptaPlannerTestResourceTest` integration test. 

As for the test issue, I'm going to do some more digging, I'll ping you if I have further questions. Thanks! :)

Comment (cuijulian) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: @geoand @ge0ffrey Fixed by modifying the `config` using `withAdapters()`, the integration test works with expected output, for both non native and native mode.

Comment (geoand) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: Excellent!

**Comment by geoand** [comment]
It's probably best to add to add `quarkus-optaplanner-jsonb-deployment` to `bom/deployment/pom.xml`. If you could add the optaplanner-jackson there as well, that would great (it seems to have been forgotten).

File: extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java
Comment (ge0ffrey) [inline extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: This exposes the list of score implementations in quarkus, which is a maintenance problem in the long term. We regularly add score implementations upstream.
@cuijulian Can you open a jira to "add OptaPlannerJsonbConfig.getScoreAdaptorList() in optaplanner-persistence-jaxb" and add a comment here, something like "// TODO replace by <jira-url> once released to avoid hardcoding the score implementations" 
Then this PR is mergable. Once that jira is fixed and optaplanner is released, we send in a PR to quarkus to fix the TODO.

Comment (cuijulian) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: Jira: https://issues.redhat.com/browse/PLANNER-1922. Comment added, changes pushed in latest commit.

**Comment by cuijulian** [comment]
> It's probably best to add to add `quarkus-optaplanner-jsonb-deployment` to `bom/deployment/pom.xml`. If you could add the optaplanner-jackson there as well, that would great (it seems to have been forgotten).

@geoand Done, nice catch! Once this is good to merge I will undraft this PR when OptaPlanner version `7.35.0.Final` is released and bump the version in `quarkus/bom`.

**Comment by geoand** [comment]
It's good to be undrafted from my POV

**Comment by ge0ffrey** [comment]
Good to merge from my POV

**Comment by geoand** [comment]
I'll take a last look tomorrow morning and if all is well I'll merge it

**Comment by cuijulian** [comment]
> I'll take a last look tomorrow morning and if all is well I'll merge it

@geoand Thank you! Note that this cannot be merged unless `optaplanner-bom` version `7.35.0.Final` is out, since there are dependencies that are not yet available on http://mvnrepository.com/. 7.35.Final should be out soon, so once it's out I'll ping you and undraft this PR for merge.

**Comment by geoand** [comment]
Oh, thanks for the info!

Then when all pieces are in place, please ping me for a final review.

File: extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java
Comment (ge0ffrey) [inline extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: @cuijulian (very soft suggestion - as we need to wait for optaplanner 7.35 to go out anyway): would you mind adding a hint in this TODO when it will be possible to fix it? For example: "when upgrading to optaplanner 7.36" (that TODO fix didn't make 7.35 which is being released). Before it was merged, I'd use the Jira url.
Also it's unclear which class has "getScoreAdaptorList()", making it hard for others your yourself in 6 months to figure out this TODO easily.

Comment (cuijulian) [reply extensions/optaplanner-jsonb/runtime/src/main/java/io/quarkus/optaplanner/jsonb/OptaPlannerJsonbConfigCustomizer.java]: Good point. I may as well also bump the `optaplanner-bom` version to 7.35 in this PR before it get merged.

**Comment by cuijulian** [comment]
@geoand OptaPlanner 7.35.0.Final is out! Should be good to merge with your approval. Thanks for your help! :)

**Review by geoand — APPROVED** [review]
Thanks a lot for this!

Let's also get a final approval from @ge0ffrey before merging

File: extensions/optaplanner-jsonb/deployment/pom.xml
Comment (gsmet) [inline extensions/optaplanner-jsonb/deployment/pom.xml:26]: Shouldn't this be a dependency on `quarkus-jsonb-deployment` instead?

File: extensions/optaplanner-jsonb/runtime/pom.xml
Comment (gsmet) [inline extensions/optaplanner-jsonb/runtime/pom.xml:27]: If you have the optaplanner dependency here, you need the deployment counterpart in the deployment pom.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added two comments regarding some weird things in the dependencies.

File: extensions/optaplanner-jsonb/deployment/pom.xml
Comment (cuijulian) [reply extensions/optaplanner-jsonb/deployment/pom.xml:26]: The `quarkus-jsonb-spi` dependency is needed in the deployment module according to the [docs](https://quarkus.io/guides/writing-extensions#customizing-json-b). But I am missing the `quarkus-jsonb-deployment` dep here, fixing in next commit.

Comment (gsmet) [reply extensions/optaplanner-jsonb/deployment/pom.xml:26]: Well, the SPI should come with the deployment module.

Comment (cuijulian) [reply extensions/optaplanner-jsonb/deployment/pom.xml:30]: Good point. Sorry, didn't check :/ Fixing in next push.

Perhaps the docs should instead say to add `quarkus-jsonb-deployment` rather than `quarkus-jsonb-spi`? Same goes for `quarkus-jackson-deployment`. I can make a PR for that.

Comment (gsmet) [reply extensions/optaplanner-jsonb/deployment/pom.xml:30]: Well, not really. You want to use the SPI if your dependency to Jackson/JSON-B is optional (which is often the case).

Here you really want a hard dependency so you don't have to use the SPI: you need the whole extension.

**Review by ge0ffrey — APPROVED** [review]

**Comment by cuijulian** [comment]
@gsmet Changes pushed. Please let me know if there are any other concerns. Thanks!

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Merged, thanks! (and sorry for the delay...)

**Pull request merged**
